android/fakedapp/src/main/java/com/solana/mobilewalletadapter/fakedapp/MainViewModel.kt

```diff
@@ -19,20 +18,20 @@ import kotlinx.coroutines.*
 import kotlinx.coroutines.flow.MutableStateFlow
 import kotlinx.coroutines.flow.asStateFlow
 import kotlinx.coroutines.flow.update
-import kotlinx.coroutines.sync.Mutex
-import kotlinx.coroutines.sync.withLock
+import kotlinx.coroutines.sync.Semaphore
+import kotlinx.coroutines.sync.withPermit
 import java.io.IOException
 import java.util.concurrent.CancellationException
 import java.util.concurrent.ExecutionException
 import java.util.concurrent.TimeUnit
 import java.util.concurrent.TimeoutException
 import kotlin.random.Random
 
-class MainViewModel(application: Application) : AndroidViewModel(application) {
+class MainViewModel : ViewModel() {
     private val _uiState = MutableStateFlow(UiState())
     val uiState = _uiState.asStateFlow()
 
-    private val mobileWalletAdapterClientMutex = Mutex() // allow only a single MWA connection at a time
+    private val mobileWalletAdapterClientSem = Semaphore(1) // allow only a single MWA connection at a time
```

Comment:
I played with a few other options here - using a channel, using a special coroutine scope with a known single-threaded context, and a Semaphore. Ultimately, I think that the Semaphore is the appropriate abstraction. The concept of permits better matches what we are trying to express than a Mutex (which implies a concrete resource).

Reply:
So you may not be wrong, but again, this seems to be overkill. An easier way would be to just enforce this at the UI level. Have the ViewModel ViewState have an "isWorking" property that gets set to true at the beginning of each operation and set to false when done. The UI could check that value before kicking off any operation.

I guess what I'm thinking through is what happens if the user taps two buttons in rapid succession? Does the semaphore just keep the second operation from starting? Does it "magically" queue the second op? The semaphore/Mutex actually makes it unclear what actually will happen. Enforcing at the ViewState/UI level is clear about what will and won't happen.

Reply:
I envision this as a situation that will probably be abstracted away by the convenience wrappers, which are better positioned to have an opinion on how multiple consumers will be handled.

In terms of the technical details, it acts like a regular Semaphore would, but substitute suspending for blocking. For e.g., if a second consumer attempts to acquire a permit when none are available, the coroutine context will be suspended until one is available. Since it's a coroutine sync primitive, it respects cancelability.

Reply:
Again, my main concern is the type of code we're putting out there as the "right" way to do things. I don't necessarily want people to think they have to do complicated things just to get baseline functionality.

That said, if you're cool with the fake dApp being rewritten to use one of the convenience wrappers once we create them, I'm cool with this.

Reply:
Yeah I agree:
- Complex code is often needed to handle complex problems and situations. This code looks good to me, funnily it reminds me exactly of the logic for "Downloads" which has similar problems to solve I guess.
- Hiding this complex code behind easier-to-use or easier-to-understand APIs might be the next priority, also by making these new APIs the default implementation in the example apps (fake DApp etc...).

I guess one question I would have would be: in addition to the Coroutines and Rx APIs, are we thinking about having a lightweight (I.E. without important external dependencies) wrapper that simplifies the implementation for devs? Like using an Executor, or something like that.